lockfile/src/cyclonedx.rs

```diff
@@ -0,0 +1,313 @@
+    }
+
+    fn purl(&self) -> Option<&str> {
+        self.purl.as_deref()
+    }
+
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized,
+    {
+        self.components.as_deref()
+    }
+}
+
+// Filter components based on their type and scope.
+fn filter_components<T: Component>(components: &[T]) -> impl Iterator<Item = &'_ T> {
+    components
+        .iter()
+        .filter(|&comp| {
+            let type_check = comp.component_type() == "application"
+                || comp.component_type() == "framework"
+                || comp.component_type() == "library";
+
+            // Check if the scope is "required" or not specified (required)
+            let scope_check = match comp.scope() {
+                Some(scope) => scope == "required",
+                None => true,
+            };
+
+            type_check && scope_check
+        })
+        .flat_map(|comp| {
+            let nested_iter = match comp.components() {
+                Some(nested) => filter_components(nested).collect::<Vec<_>>(),
+                None => Vec::new(),
+            };
+            std::iter::once(comp).chain(nested_iter.into_iter())
+        })
+}
+
+// Convert a component's Package URL (PURL) into a Package object.
+fn from_purl<T: Component>(component: &T) -> anyhow::Result<Package> {
+    let purl_str = component
+        .purl()
+        .ok_or_else(|| anyhow!("Missing purl for {}:{}", component.name(), component.version()))?;
+    let purl = GenericPurl::<String>::from_str(purl_str)?;
+    let package_type = PackageType::from_str(purl.package_type()).map_err(|_| UnknownEcosystem)?;
+
+    // Determine the package name based on its type and potential namespace.
+    let name = match (package_type, purl.namespace()) {
+        (PackageType::Maven, Some(ns)) => format!("{}:{}", ns, purl.name()),
+        (PackageType::Npm | PackageType::Golang, Some(ns)) => format!("{}/{}", ns, purl.name()),
+        _ => purl.name().into(),
+    };
+
+    // Extract the package version
+    let pkg_version = purl
+        .version()
+        .ok_or(&component.version())
+        .map_err(|_| anyhow!("No version found for `{}`", name))?;
+
+    // Use the qualifiers from the PURL to determine the version details.
+    let version = purl
+        .qualifiers()
+        .iter()
+        .find_map(|(key, value)| match key.as_ref() {
+            "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                version: pkg_version.into(),
+                registry: value.to_string(),
+            })),
+            "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+            "vcs_url" => {
+                if value.starts_with("git+") {
+                    Some(PackageVersion::Git(value.to_string()))
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        })
+        .unwrap_or(PackageVersion::FirstParty(pkg_version.into()));
```

Comment:
This also seems like a concept that applies to more than just CycloneDX. I'm not sure if this was necessary in spdx already, but probably worth it to at least extract it into a method so it can be moved into a more generic position if we ever need to reuse it.